File: advent2017/extras/cxx/src/days/day21.cpp
// Architecture Notes:
// - This translation unit is documented for codebase-wide recoverability.
// - Solver stages are explicit to simplify verification and benchmarking.
// - Low-level optimizations are annotated with rationale and invariants.

#include <algorithm>
#include <chrono>
#include <cstdint>
#include <fstream>
#include <functional>
#include <iostream>
#include <map>
#include <regex>
#include <stdexcept>
#include <string>
#include <unordered_map>
#include <vector>
#include <openssl/sha.h>

namespace {
constexpr const char* EXPECTED_SHA = "759a25acf919be68478e4d20d3856f488ff79325d0954d8ca5c89cecc2fd8287";
constexpr const char* EXPECTED_PART1 = "139";
constexpr const char* EXPECTED_PART2 = "1857134";

std::string resolve_input(const std::string& provided) {
    if (!provided.empty()) return provided;
    std::vector<std::string> cands = {
        "advent2017/Day21/d21_input.txt",
        "Day21/d21_input.txt",
        "../Day21/d21_input.txt",
        "../../Day21/d21_input.txt",
    };
    for (const auto& c : cands) {
        std::ifstream f(c);
        if (f.good()) return c;
    }
    throw std::runtime_error("input not found");
}

std::string sha256_file(const std::string& path) {
    std::ifstream f(path, std::ios::binary);
    if (!f) throw std::runtime_error("read fail");
    std::vector<unsigned char> b((std::istreambuf_iterator<char>(f)), {});
    unsigned char out[SHA256_DIGEST_LENGTH];
    SHA256(b.data(), b.size(), out);
    static const char* HEX = "0123456789abcdef";
    std::string s;
    s.reserve(64);
    for (unsigned char v : out) {
        s.push_back(HEX[v >> 4]);
        s.push_back(HEX[v & 15]);
    }
    return s;
}

std::vector<std::string> rotate(const std::vector<std::string>& p) {
    int n = static_cast<int>(p.size());
    std::vector<std::string> out(n, std::string(n, '.'));
    for (int c = 0; c < n; ++c) {
        for (int r = 0; r < n; ++r) out[c][r] = p[n - 1 - r][c];
    }
    return out;
}

std::vector<std::string> flip(const std::vector<std::string>& p) {
    std::vector<std::string> out = p;
    for (auto& row : out) std::reverse(row.begin(), row.end());
    return out;
}

std::string join_pat(const std::vector<std::string>& p) {
    std::string s;
    for (size_t i = 0; i < p.size(); ++i) {
        if (i) s.push_back('/');
        s += p[i];
    }
    return s;
}

std::string canonical(std::vector<std::string> p) {
    std::vector<std::string> vars;
    vars.reserve(8);
    for (int i = 0; i < 4; ++i) {
        vars.push_back(join_pat(p));
        vars.push_back(join_pat(flip(p)));
        p = rotate(p);
    }
    std::sort(vars.begin(), vars.end());
    return vars[0];
}

std::unordered_map<std::string, std::vector<std::string>> parse_rules(const std::string& path) {
    std::ifstream f(path);
    if (!f) throw std::runtime_error("open fail");
    std::unordered_map<std::string, std::vector<std::string>> rules;
    std::string line;
    while (std::getline(f, line)) {
        if (line.empty()) continue;
        auto pos = line.find(" => ");
        auto lhs = line.substr(0, pos);
        auto rhs = line.substr(pos + 4);
        std::vector<std::string> in, out;
        size_t s = 0, e = 0;
        while ((e = lhs.find('/', s)) != std::string::npos) {
            in.push_back(lhs.substr(s, e - s));
            s = e + 1;
        }
        in.push_back(lhs.substr(s));
        s = 0;
        while ((e = rhs.find('/', s)) != std::string::npos) {
            out.push_back(rhs.substr(s, e - s));
            s = e + 1;
        }
        out.push_back(rhs.substr(s));
        rules[canonical(in)] = out;
    }
    return rules;
}

std::vector<std::string> enhance(const std::vector<std::string>& grid,
                                 const std::unordered_map<std::string, std::vector<std::string>>& rules) {
    int n = static_cast<int>(grid.size());
    int bs = (n % 2 == 0) ? 2 : 3;
    int os = bs + 1;
    int cnt = n / bs;
    int new_n = cnt * os;
    std::vector<std::string> out(new_n, std::string(new_n, '.'));
    for (int br = 0; br < cnt; ++br) {
        for (int bc = 0; bc < cnt; ++bc) {
            std::vector<std::string> sub(bs);
            for (int r = 0; r < bs; ++r) sub[r] = grid[br * bs + r].substr(bc * bs, bs);
            const auto& rep = rules.at(canonical(sub));
            for (int r = 0; r < os; ++r)
                for (int c = 0; c < os; ++c) out[br * os + r][bc * os + c] = rep[r][c];
        }
    }
    return out;
}

int run(const std::unordered_map<std::string, std::vector<std::string>>& rules, int iters) {
    std::vector<std::string> grid = {".#.", "..#", "###"};
    for (int i = 0; i < iters; ++i) grid = enhance(grid, rules);
    int total = 0;
    for (const auto& row : grid)
        for (char ch : row)
            if (ch == '#') ++total;
    return total;
}

std::vector<std::string> split_blocks(const std::vector<std::string>& grid, int bs) {
    int n = static_cast<int>(grid.size());
    int cnt = n / bs;
    std::vector<std::string> out;
    out.reserve(cnt * cnt);
    for (int br = 0; br < cnt; ++br) {
        for (int bc = 0; bc < cnt; ++bc) {
            std::vector<std::string> sub(bs);
            for (int r = 0; r < bs; ++r) sub[r] = grid[br * bs + r].substr(bc * bs, bs);
            out.push_back(join_pat(sub));
        }
    }
    return out;
}

int popcount_key(const std::string& k) {
    return static_cast<int>(std::count(k.begin(), k.end(), '#'));
}

int run_optimized(const std::unordered_map<std::string, std::vector<std::string>>& rules, int iters) {
    if (iters <= 5 || (iters % 3) != 0) return run(rules, iters);

    std::unordered_map<std::string, std::vector<std::string>> memo_expand;
    std::unordered_map<std::string, int> memo_count;

    std::function<std::vector<std::string>(const std::string&)> expand_three_from_3 =
        [&](const std::string& k) -> std::vector<std::string> {
        auto it = memo_expand.find(k);
        if (it != memo_expand.end()) return it->second;
        std::vector<std::string> g;
        size_t s = 0, e = 0;
        while ((e = k.find('/', s)) != std::string::npos) {
            g.push_back(k.substr(s, e - s));
            s = e + 1;
        }
        g.push_back(k.substr(s));
        g = enhance(g, rules);
        g = enhance(g, rules);
        g = enhance(g, rules);
        auto out = split_blocks(g, 3);
        memo_expand[k] = out;
        return out;
    };

    std::function<int(const std::string&, int)> count_cycles =
        [&](const std::string& k, int cycles) -> int {
        std::string mk = std::to_string(cycles) + "|" + k;
        auto it = memo_count.find(mk);
        if (it != memo_count.end()) return it->second;
        int v = 0;
        if (cycles == 0) {
            v = popcount_key(k);
        } else {
            for (const auto& sub : expand_three_from_3(k)) v += count_cycles(sub, cycles - 1);
        }
        memo_count[mk] = v;
        return v;
    };

    return count_cycles(".#./..#/###", iters / 3);
}
}  // namespace

int main(int argc, char** argv) {
    int part = 0;
    std::string input;
    for (int i = 1; i < argc; ++i) {
        std::string a = argv[i];
        if (a == "--part")
            part = std::stoi(argv[++i]);
        else if (a == "--input")
            input = argv[++i];
        else
            throw std::runtime_error("unknown arg");
    }
    if (part != 1 && part != 2) throw std::runtime_error("--part 1|2 required");
    input = resolve_input(input);
    if (sha256_file(input) != EXPECTED_SHA) throw std::runtime_error("checksum mismatch");
    auto rules = parse_rules(input);

    auto t0 = std::chrono::steady_clock::now();
    std::string ans = std::to_string(run_optimized(rules, part == 1 ? 5 : 18));
    std::string expected = part == 1 ? EXPECTED_PART1 : EXPECTED_PART2;
    if (ans != expected) throw std::runtime_error("answer mismatch");
    std::cout << ans << "\n";
    auto t1 = std::chrono::steady_clock::now();
    double ms = std::chrono::duration<double, std::milli>(t1 - t0).count();
    std::cerr << "[cxx-fancy] day=21 part=" << part << " runtime_ms=" << ms << "\n";
    return 0;
}

File: advent2017/extras/rust/src/bin/day21.rs
use sha2::{Digest, Sha256};
use std::collections::HashMap;
use std::env;
use std::fs;
use std::path::Path;
use std::time::Instant;

const EXPECTED_SHA: &str = "759a25acf919be68478e4d20d3856f488ff79325d0954d8ca5c89cecc2fd8287";
const EXPECTED_PART1: &str = "139";
const EXPECTED_PART2: &str = "1857134";

fn resolve_input(provided: &str) -> String {
    if !provided.is_empty() {
        return provided.to_string();
    }
    let cands = [
        "advent2017/Day21/d21_input.txt",
        "Day21/d21_input.txt",
        "../Day21/d21_input.txt",
        "../../Day21/d21_input.txt",
    ];
    for c in cands {
        if Path::new(c).exists() {
            return c.to_string();
        }
    }
    panic!("input not found");
}

fn sha256_file(path: &str) -> String {
    let b = fs::read(path).expect("read");
    let mut h = Sha256::new();
    h.update(&b);
    hex::encode(h.finalize())
}

fn rotate(p: &[String]) -> Vec<String> {
    let n = p.len();
    let mut out = vec![String::new(); n];
    for c in 0..n {
        let mut row = String::with_capacity(n);
        for r in 0..n {
            row.push(p[n - 1 - r].as_bytes()[c] as char);
        }
        out[c] = row;
    }
    out
}

fn flip(p: &[String]) -> Vec<String> {
    p.iter()
        .map(|r| r.chars().rev().collect::<String>())
        .collect::<Vec<String>>()
}

fn canonical(p: &[String]) -> String {
    let mut cur = p.to_vec();
    let mut vars = Vec::with_capacity(8);
    for _ in 0..4 {
        vars.push(cur.join("/"));
        vars.push(flip(&cur).join("/"));
        cur = rotate(&cur);
    }
    vars.sort_unstable();
    vars[0].clone()
}

fn parse_rules(text: &str) -> HashMap<String, Vec<String>> {
    let mut rules = HashMap::new();
    for ln in text.lines() {
        let s = ln.trim();
        if s.is_empty() {
            continue;
        }
        let mut it = s.split(" => ");
        let inp = it
            .next()
            .expect("lhs")
            .split('/')
            .map(|x| x.to_string())
            .collect::<Vec<String>>();
        let out = it
            .next()
            .expect("rhs")
            .split('/')
            .map(|x| x.to_string())
            .collect::<Vec<String>>();
        rules.insert(canonical(&inp), out);
    }
    rules
}

fn enhance(grid: &[String], rules: &HashMap<String, Vec<String>>) -> Vec<String> {
    let n = grid.len();
    let bs = if n % 2 == 0 { 2 } else { 3 };
    let os = bs + 1;
    let cnt = n / bs;
    let new_n = cnt * os;
    let mut out = vec![vec!['.'; new_n]; new_n];
    for br in 0..cnt {
        for bc in 0..cnt {
            let mut sub = Vec::with_capacity(bs);
            for r in 0..bs {
                sub.push(grid[br * bs + r][bc * bs..bc * bs + bs].to_string());
            }
            let rep = rules.get(&canonical(&sub)).expect("rule");
            for r in 0..os {
                for c in 0..os {
                    out[br * os + r][bc * os + c] = rep[r].as_bytes()[c] as char;
                }
            }
        }
    }
    out.into_iter()
        .map(|row| row.into_iter().collect::<String>())
        .collect()
}

fn run_iterative(rules: &HashMap<String, Vec<String>>, iters: usize) -> usize {
    let mut grid = vec![".#.".to_string(), "..#".to_string(), "###".to_string()];
    for _ in 0..iters {
        grid = enhance(&grid, rules);
    }
    grid.iter()
        .map(|r| r.bytes().filter(|&b| b == b'#').count())
        .sum()
}

fn split_blocks(grid: &[String], bs: usize) -> Vec<String> {
    let n = grid.len();
    let cnt = n / bs;
    let mut out = Vec::with_capacity(cnt * cnt);
    for br in 0..cnt {
        for bc in 0..cnt {
            let mut sub = Vec::with_capacity(bs);
            for r in 0..bs {
                sub.push(grid[br * bs + r][bc * bs..bc * bs + bs].to_string());
            }
            out.push(sub.join("/"));
        }
    }
    out
}

fn popcount_key(k: &str) -> usize {
    k.bytes().filter(|&b| b == b'#').count()
}

fn run_optimized(rules: &HashMap<String, Vec<String>>, iters: usize) -> usize {
    if iters <= 5 || iters % 3 != 0 {
        return run_iterative(rules, iters);
    }
    let mut memo_expand: HashMap<String, Vec<String>> = HashMap::new();
    let mut memo_count: HashMap<(String, usize), usize> = HashMap::new();

    fn expand_three_from_3(
        k: &str,
        rules: &HashMap<String, Vec<String>>,
        memo_expand: &mut HashMap<String, Vec<String>>,
    ) -> Vec<String> {
        if let Some(v) = memo_expand.get(k) {
            return v.clone();
        }
        let mut g = k.split('/').map(|s| s.to_string()).collect::<Vec<String>>();
        g = enhance(&g, rules);
        g = enhance(&g, rules);
        g = enhance(&g, rules);
        let out = split_blocks(&g, 3);
        memo_expand.insert(k.to_string(), out.clone());
        out
    }

    fn count_cycles(
        k: &str,
        cycles: usize,
        rules: &HashMap<String, Vec<String>>,
        memo_expand: &mut HashMap<String, Vec<String>>,
        memo_count: &mut HashMap<(String, usize), usize>,
    ) -> usize {
        if let Some(v) = memo_count.get(&(k.to_string(), cycles)) {
            return *v;
        }
        let v = if cycles == 0 {
            popcount_key(k)
        } else {
            expand_three_from_3(k, rules, memo_expand)
                .iter()
                .map(|sub| count_cycles(sub, cycles - 1, rules, memo_expand, memo_count))
                .sum()
        };
        memo_count.insert((k.to_string(), cycles), v);
        v
    }

    count_cycles(
        ".#./..#/###",
        iters / 3,
        rules,
        &mut memo_expand,
        &mut memo_count,
    )
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let mut part = 0;
    let mut input = String::new();
    let mut i = 1;
    while i < args.len() {
        match args[i].as_str() {
            "--part" => {
                part = args[i + 1].parse().unwrap();
                i += 2;
            }
            "--input" => {
                input = args[i + 1].clone();
                i += 2;
            }
            _ => panic!("unknown arg"),
        }
    }
    if part != 1 && part != 2 {
        panic!("--part 1|2 required");
    }
    let in_path = resolve_input(&input);
    if sha256_file(&in_path) != EXPECTED_SHA {
        panic!("checksum mismatch");
    }
    let rules = parse_rules(&fs::read_to_string(&in_path).expect("read_to_string"));

    let t0 = Instant::now();
    let ans = if part == 1 {
        run_optimized(&rules, 5).to_string()
    } else {
        run_optimized(&rules, 18).to_string()
    };
    let expected = if part == 1 {
        EXPECTED_PART1
    } else {
        EXPECTED_PART2
    };
    if ans != expected {
        panic!("answer mismatch");
    }
    println!("{}", ans);
    let ms = t0.elapsed().as_nanos() as f64 / 1e6;
    eprintln!("[rust-fancy] day=21 part={} runtime_ms={:.3}", part, ms);
}
